src/v/cluster/self_test_frontend.cc

```diff
@@ -0,0 +1,263 @@
+    });
+    co_return enrich(v);
+}
+
+ss::future<self_test_frontend::local_test_state>
+self_test_frontend::remote_invoke::get_status() {
+    auto v = co_await send_self_test_rpc([](self_test_rpc_client_protocol c) {
+        return c.get_status(
+          empty_request{}, rpc::client_opts(raft::clock_type::now() + 5s));
+    });
+    co_return enrich(v);
+}
+
+/// Implementation of methods in self_test_frontend
+
+self_test_frontend::self_test_frontend(
+  model::node_id self,
+  ss::sharded<cluster::members_table>& members,
+  ss::sharded<cluster::self_test_backend>& self_be,
+  ss::sharded<rpc::connection_cache>& connections)
+  : _self(self)
+  , _state({})
+  , _previous_successful_run({})
+  , _members(members)
+  , _self_be(self_be)
+  , _connections(connections) {}
+
+ss::future<> self_test_frontend::start() { return ss::now(); }
+
+ss::future<> self_test_frontend::stop() { co_await _gate.close(); }
+
+ss::future<uuid_t>
+self_test_frontend::start_test(diskcheck_opts dto, netcheck_opts nto) {
+    auto gate_holder = _gate.hold();
+    const auto stopped_results = co_await stop_test();
+    if (!stopped_results.finished()) {
+        /// hmm what to do? maybe throw? continue with less participants
+        co_return uuid_t{}; // TODO: fix
+    }
+
+    /// Invoke command to start test on all nodes, using the same test id
+    const auto id = uuid_t::create();
+    start_test_request req{.id = id, .dto = dto, .nto = nto};
+    co_await invoke_on_all_nodes(
+      [req](auto handle) { return handle->start_test(req); });
+    /// TODO: Throw if something bad occurred
+    co_return id;
+}
+
+ss::future<self_test_frontend::global_test_state>
+self_test_frontend::stop_test() {
+    auto gate_holder = _gate.hold();
+    int8_t retries = 3;
+    auto test_state = co_await status();
+    while (!test_state.finished() && retries-- > 0) {
+        /// invoke_on_all will filter brokers that have shut down cleanly from
+        /// its participants list, retries (if occur) will only occur on
+        /// previously unreachable nodes
+        test_state = co_await invoke_on_all_nodes(
+          [](auto handle) { return handle->stop_test(); });
+        if (!test_state.finished()) {
+            vlog(
+              clusterlog.warn,
+              "Failed to sucessfully stop all outstanding jobs on all "
+              "participants: {} , trying again, retries: {}",
+              test_state,
+              retries);
+        }
+        co_await ss::sleep(1s);
+    }
+    co_return test_state;
+}
+
+ss::future<self_test_frontend::global_test_state> self_test_frontend::status() {
+    return invoke_on_all_nodes(
+      [](auto handle) { return handle->get_status(); });
+}
+
+std::vector<model::node_id>
+self_test_frontend::get_active_participants() const {
+    auto participants = _state.active_participant_ids();
+    auto brokers = participants.empty() ? _members.local().node_ids()
```

Comment:
This seems a bit nuanced and probably deserves a comment. If there are no more pending participants, that means the previous test is done, right? Which is why we return all nodes? I find it a bit confusing because I think there's an assumption that after a test finishes, the next call will be `start_test`, otherwise we end up sending RPCs fruitlessly for a test that didn't finish?

Reply:
Return all nodes because its assumed a new test run is to start. Its up to the caller to check the current status of the world before calling the invoke_on_all method.

Reply:
commenting and moving into anon-lambda to remove confusion